app/src/main/java/de/westnordost/streetcomplete/screens/user/profile/ProfileFragment.kt

```diff
@@ -114,33 +52,60 @@ class ProfileFragment : Fragment(R.layout.fragment_profile) {
         binding.currentWeekLocalRankText.background = LaurelWreathDrawable(resources)
         binding.currentWeekGlobalRankText.background = LaurelWreathDrawable(resources)
 
-        binding.logoutButton.setOnClickListener {
-            userLoginStatusController.logOut()
-        }
+        binding.logoutButton.setOnClickListener { viewModel.logOutUser() }
         binding.profileButton.setOnClickListener {
-            openUri("https://www.openstreetmap.org/user/" + userDataSource.userName)
+            openUri("https://www.openstreetmap.org/user/" + viewModel.userName.value)
         }
-    }
-
-    override fun onStart() {
-        super.onStart()
 
-        viewLifecycleScope.launch {
-            userDataSource.addListener(userListener)
-            userUpdater.addUserAvatarListener(userAvatarListener)
-            statisticsSource.addListener(statisticsListener)
-            unsyncedChangesCountSource.addListener(unsyncedChangesCountListener)
-            achievementsSource.addListener(achievementsListener)
-
-            updateUserName()
-            updateAvatar()
-            updateEditCountTexts()
-            updateUnpublishedEditsText()
-            updateDaysActiveText()
-            updateGlobalRankTexts()
-            updateLocalRankTexts()
-            updateAchievementLevelsText()
-            updateDatesActiveView()
+        observe(viewModel.userName) { name ->
+            binding.userNameTextView.text = name
+        }
+        observe(viewModel.userAvatarFile) { file ->
+            val avatar = if (file.exists()) BitmapFactory.decodeFile(file.path) else anonAvatar
```

Comment:
IO Business code in UI, let VM return a `Bitmap?` which is null if the decoding fails, then default to anon avatar in UI.

Reply:
`Bitmap` is Android specific, though, and my aim would be to keep ViewModels free of any platform specific code. Compose Multiplatform maybe brings their own `Bitmap` (like) + File system access classes (probably not), but then the ViewModels depend on a specific UI framework.
I'd like to avoid that. Anyway, as with the other thing with Intent/Activity, I tend towards postponing thinking about that.